core/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/entity/decorator/FilteringEntityByEntitySelector.java

```diff
@@ -0,0 +1,411 @@
+        } else {
+            throw new IllegalStateException("The selector (%s) does not support a ListIterator with randomSelection (%s)."
+                    .formatted(this, randomSelection));
+        }
+    }
+
+    @Override
+    public ListIterator<Object> listIterator(int index) {
+        if (!randomSelection) {
+            return new OriginalFilteringValueRangeIterator<>(this::selectReplayedEntity,
+                    childEntitySelector.listIterator(index), basicVariableDescriptors, valueRangeManager);
+        } else {
+            throw new IllegalStateException("The selector (%s) does not support a ListIterator with randomSelection (%s)."
+                    .formatted(this, randomSelection));
+        }
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other instanceof FilteringEntityByEntitySelector<?> that
+                && Objects.equals(childEntitySelector, that.childEntitySelector)
+                && Objects.equals(replayingEntitySelector, that.replayingEntitySelector);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(childEntitySelector, replayingEntitySelector);
+    }
+
+    private abstract static class AbstractFilteringValueRangeIterator<Solution_> implements Iterator<Object> {
+        private final Supplier<Object> upcomingEntitySupplier;
+        private final BasicVariableDescriptor<Solution_>[] basicVariableDescriptors;
+        private final ValueRangeManager<Solution_> valueRangeManager;
+        private boolean initialized = false;
+        private Object replayedEntity;
+
+        private AbstractFilteringValueRangeIterator(Supplier<Object> upcomingEntitySupplier,
+                BasicVariableDescriptor<Solution_>[] basicVariableDescriptors, ValueRangeManager<Solution_> valueRangeManager) {
+            this.upcomingEntitySupplier = upcomingEntitySupplier;
+            this.basicVariableDescriptors = basicVariableDescriptors;
+            this.valueRangeManager = valueRangeManager;
+        }
+
+        void initialize() {
+            if (initialized) {
+                return;
+            }
+            checkReplayedEntity();
+            initialized = true;
+        }
+
+        void checkReplayedEntity() {
+            var updatedReplayedEntity = upcomingEntitySupplier.get();
+            if (replayedEntity == null || replayedEntity != updatedReplayedEntity) {
+                replayedEntity = updatedReplayedEntity;
+            }
+        }
+
+        Object currentReplayedEntity() {
+            return replayedEntity;
+        }
+
+        /**
+         * The other entity is reachable if it accepts all assigned values from the replayed entity, and vice versa.
+         */
+        boolean isReachable(Object otherEntity) {
+            if (replayedEntity == otherEntity) {
+                // Same entity cannot be swapped
+                return false;
+            }
+
+            if (basicVariableDescriptors.length == 1) {
```

Comment:
Is this a measurable performance improvement?
If not, I'd rather use the loop always, to simplify code.